### PR TITLE
Add logback encoder support and use UTF-8 encoding explicitly

### PR DESCRIPTION
**Pull request by rillgen**
We're using this library to push our logs using logback, and would like to send them encoded in JSON format so that we can push custom logging attributes and use them for filtering.

There are logback encoders that do this such as elastic's `LogstashEncoder` but unfortunately the r7insight library lacks support for these as when an `Encoder` is defined, it assumes that it is a `PatternLayoutEncoder`. This patch aims to provide generic support for any `Encoder`.

### CODE REVIEW

**Comment by sbialkowski-r7** [comment]
Hi @rillgen, seems that this repo uses spaces, not tabs, this causes a massive diff, would you be able to reformat?

**Comment by rillgen** [comment]
> Hi @rillgen, seems that this repo uses spaces, not tabs, this causes a massive diff, would you be able to reformat?

Hi @sbialkowski-r7 thanks for pointing that out, I reformatted the code using spaces.

File: src/test/java/com/rapid7/util/SocketChannelReceiver.java [hunk 1]
```diff
@@ -119,7 +120,7 @@ private void listenForMessages(Socket socket) throws IOException {
             if (messageLength == -1) {
                 continue;
             }
-            String[] messages = new String(Arrays.copyOf(buffer, messageLength)).split("\\r?\\n");
+            String[] messages = new String(Arrays.copyOf(buffer, messageLength), UTF_8).split("\\r?\\n");
```
Comment (sbialkowski-r7) [inline src/test/java/com/rapid7/util/SocketChannelReceiver.java]: can this throw a `UnsupportedEncodingException`?

Comment (rillgen) [reply src/test/java/com/rapid7/util/SocketChannelReceiver.java]: Only the `String` constructor that receives the `charsetName` parameter as a `String` throws an `UnsupportedEncodingException` but here I'm using the one which receives a `Charset` from `StandardCharsets` which doesn't.

File: src/main/java/com/rapid7/logback/LogentriesAppender.java [hunk 2]
```diff
@@ -240,7 +249,13 @@ public void setLayout(Layout<ILoggingEvent> layout) {
     @Override
     protected void append(ILoggingEvent event) {
         // Render the event according to layout
-        String formattedEvent = layout.doLayout(event);
+        byte[] encodedEvent = encoder.encode(event);
+        String formattedEvent;
+        try {
+            formattedEvent = new String(encodedEvent, "UTF-8");
```
Comment (jcolgan-r7) [inline src/main/java/com/rapid7/logback/LogentriesAppender.java]: Can we use this? I believe it removes the need for the try-catch.

```suggestion
            formattedEvent = new String(encodedEvent, StandardCharsets.UTF_8);
```

Comment (rillgen) [reply src/main/java/com/rapid7/logback/LogentriesAppender.java]: 👍

File: src/main/java/com/rapid7/logback/LogentriesAppender.java [hunk 3]
```diff
@@ -171,18 +176,22 @@ public void setLogID(String logID) {
...
+        this.encoder = encoder;
     }
 
     @Override
     public void start() {
-        if (layout == null) {
-            layout = buildLayout();
+        if (encoder == null) {
+            layout = layout == null ? buildLayout() : layout;
+            LayoutWrappingEncoder lwe = new LayoutWrappingEncoder<ILoggingEvent>();
```
Comment (jcolgan-r7) [inline src/main/java/com/rapid7/logback/LogentriesAppender.java]: Probably should be in this form?

```suggestion
            LayoutWrappingEncoder<ILoggingEvent> lwe = new LayoutWrappingEncoder<>();
```

**Comment by sbialkowski-r7** [comment]
@rillgen Thanks so much for contributing to our repo, we really appreciate it!

**Pull request merged**